src/zlib.zig

```diff
@@ -901,3 +945,277 @@ pub const ZlibCompressorArrayList = struct {
+
+        _ = deflateEnd(&this.state);
+        if (this.err != .StreamEnd and this.finishFlush == .Finish) return error.ZlibError10;
+    }
+};
+
+pub const ZlibDecompressorStreaming = struct {
+    mode: NodeMode,
+    state: z_stream = std.mem.zeroes(z_stream),
+    chunkSize: c_uint,
+    next_expected_header_byte: ?[*]const u8 = null,
+    gzip_id_bytes_read: u16 = 0,
+    flush: FlushValue = .NoFlush,
+    finishFlush: FlushValue = .Finish,
+    fullFlush: FlushValue = .FullFlush,
+    windowBits: c_int,
+    dictionary: []const u8,
+    err: ReturnCode = .Ok,
+    err_msg: ?[*:0]const u8 = null,
+    do_inflate_loop: bool = true,
+
+    pub fn init(this: *ZlibDecompressorStreaming) !void {
+        const ret_code = inflateInit2_(&this.state, this.windowBits, zlibVersion(), @sizeOf(z_stream));
+        if (ret_code != .Ok) return error.ZlibError1;
+
+        this.setDictionary() catch {};
+        this.err_msg = null;
+        return;
+    }
+
+    fn setDictionary(this: *ZlibDecompressorStreaming) !void {
+        this.err = .Ok;
+        const dictionary = this.dictionary;
+        switch (this.mode) {
+            .DEFLATE, .DEFLATERAW => {
+                if (dictionary.len > 0) this.err = deflateSetDictionary(&this.state, dictionary.ptr, @intCast(dictionary.len));
+            },
+            .INFLATERAW => {
+                if (dictionary.len > 0) this.err = inflateSetDictionary(&this.state, dictionary.ptr, @intCast(dictionary.len));
+            },
+            else => {},
+        }
+        if (this.err != .Ok) {
+            return this.error_for_message("Failed to set dictionary");
+        }
+    }
+
+    fn error_for_message(this: *ZlibDecompressorStreaming, default: [*:0]const u8) error{ZlibError} {
+        var message = default;
+        if (this.state.err_msg) |msg| message = msg;
+        this.err_msg = message;
+        return error.ZlibError;
+    }
+
+    pub fn writeAll(this: *ZlibDecompressorStreaming, bytes: []const u8, output: *std.ArrayListUnmanaged(u8), process_all_input: bool) !void {
+        var index: usize = 0;
+        while (index != bytes.len) {
+            index += try this.write(bytes[index..], output, process_all_input);
+        }
+    }
+
+    fn write(this: *ZlibDecompressorStreaming, bytes: []const u8, output: *std.ArrayListUnmanaged(u8), process_all_input: bool) !usize {
+        const state = &this.state;
+        state.next_in = bytes.ptr;
+        state.avail_in = @intCast(bytes.len);
+        if (state.avail_in == 0) state.next_in = null;
+
+        // UNZIP mode allows the input to be either gzip or deflate data and we do a two-byte header detection in order to disambiguate.
+        // the ordering of this logic is a bit abstract because we dont know ahead of time how large 'bytes' will be.
+        // additionally, if the first byte is "correct" but the second is not, we don't want to lose it from it being consumed.
+        if (this.mode == .UNZIP) {
```

Comment:
I think it'd be good to link to the code this was ported from so that we have a reference implementation

https://github.com/nodejs/node/blob/2eff28fb7a93d3f672f80b582f664a7c701569fb/src/node_zlib.cc#L773